### PR TITLE
feat(aws/cloudformation): Add task to deploy a cloudformation stack

### PR DESCRIPTION
**Pull request by xavileon**
This task is responsible to deploy a cloudformation stack. It receives a
description of the cloudformation stack containing:
* the name of the stack
* the region where it's deployed
* the template of the cloudformation stack
* and the parameters used for the template

This will create the stack process in AWS and return the stackId as a
result of the task, so Orca can followup on its progress.

This PR partly addresses spinnaker/spinnaker#736 and depends on https://github.com/spinnaker/clouddriver/pull/3278

Signed-off-by: Xavi León <xavi.leon@schibsted.com>

### CODE REVIEW

**Comment by gardleopard** [comment]
@spinnaker/amazon PTAL

**Comment by xavileon** [comment]
@spinnaker/reviewers PTAL

**Comment by xavileon** [comment]
@spinnaker/reviewers PTAL

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+public class DeployCloudFormationAtomicOperation implements AtomicOperation<Map> {
+
+  private static final String BASE_PHASE = "DEPLOY_CLOUDFORMATION";
+
+
+  @Autowired
+  AmazonClientProvider amazonClientProvider;
+
+  @Autowired
+  private ObjectMapper objectMapper;
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java]: I don't think this is going to get autowired? Are you sure it does?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  public Map operate(List priorOutputs) {
+    Task task = TaskRepository.threadLocalTask.get();
+    task.updateStatus(BASE_PHASE, "Configuring cloudformation");
+    AmazonCloudFormation amazonCloudFormation = amazonClientProvider.getAmazonCloudFormation(
+      description.getCredentials(), description.getRegion());
+    String template;
+    try {
+      template = objectMapper.writeValueAsString(description.getTemplateBody());
+    } catch (JsonProcessingException e) {
+      throw new IllegalArgumentException("Could not deserialize template body", e);
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java]: We're serializing here.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  private String createStack(AmazonCloudFormation amazonCloudFormation, String template, List<Parameter> parameters) {
+    Task task = TaskRepository.threadLocalTask.get();
+    task.updateStatus(BASE_PHASE, "Preparing cloudformation");
+    CreateStackRequest createStackRequest = new CreateStackRequest()
+      .withStackName(description.getStackName())
+      .withParameters(parameters)
+      .withTemplateBody(template);
+    task.updateStatus(BASE_PHASE, "Uploading cloudformation");
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java]: I'd recommend changing these to say "[blah blah] CloudFormation Stack" instead.

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java [hunk 4]
```diff
@@ -94,4 +94,8 @@
   // Service operations
   public static final String DEPLOY_SERVICE = "deployService";
   public static final String DESTROY_SERVICE = "destroyService";
+
+  // CloudFormation operations
+  public static final String DEPLOY_CLOUDFORMATION = "deployCloudFormation";
```
Comment (robzienert) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: `DEPLOY_CLOUDFORMATION_STACK` imo.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java (hunk 1 above)
Comment (xavileon) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java]: Yes, it gets autowired 👍 configured in https://github.com/spinnaker/clouddriver/blob/master/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/config/AwsConfiguration.groovy#L122

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java (hunk 3 above)
Comment (xavileon) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java]: good point, updated 👍

**Comment by xavileon** [comment]
@robzienert Thanks for looking into it 👍 Addressed your comments in https://github.com/spinnaker/clouddriver/pull/3280/commits/f6857c526c9659f5d10e4b817ae5a156fcd2891e

**Comment by xavileon** [comment]
hey @robzienert addressed your feedback, is it ok to merge?

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java [hunk 5]
```diff
@@ -94,4 +94,8 @@
   // Service operations
   public static final String DEPLOY_SERVICE = "deployService";
   public static final String DESTROY_SERVICE = "destroyService";
+
+  // CloudFormation operations
+  public static final String DEPLOY_CLOUDFORMATION_STACK = "deployCloudFormation";
```
[the remark this answers is not in the document]
Comment (xavileon) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: Sorry, I'm still getting familiar with clouddriver 🙂 What do you mean? Could you point me to a task/operation that is not an AtomicOperation and could be an example to follow? If that suits our need I'm happy to make the necessary changes.

[the remark this answers is not in the document]
Comment (ajordens) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java]: I can circle around on this post-merge ... let's get it down for now.

**Pull request merged**